docs/serialization.md

```diff
@@ -1,10 +1,58 @@
 # Serialization
 
-YARP ships with the ability to serialize a syntax tree to a single string. The string can then be deserialized back into a syntax tree using a language other than C. This is useful for using the parsing logic in other tools without having to write a parser in that language. The syntax tree still requires a copy of the original source, as for the most part it just contains byte offsets into the source string.
+YARP ships with the ability to serialize a syntax tree to a single string.
+The string can then be deserialized back into a syntax tree using a language other than C.
+This is useful for using the parsing logic in other tools without having to write a parser in that language.
+The syntax tree still requires a copy of the original source, as for the most part it just contains byte offsets into the source string.
+
+## Types
+
+Let us define some simple types for readability.
+
+### varint
+
+A variable-length integer with the value fitting in `uint32_t` using between 1 and 5 bytes, using the [LEB128](https://en.wikipedia.org/wiki/LEB128) encoding.
+This drastically cuts down on the size of the serialized string, especially when the source file is large.
+
+### string
+
+| # bytes | field |
+| --- | --- |
+| varint | the length of the string in bytes |
+| ... | the string bytes |
+
+### location
+
+| # bytes | field |
+| --- | --- |
+| varint | byte offset into the source string where this location begins |
+| varint | length of the location in bytes in the source string |
+
+### comment
+
+The comment type is one of:
+* 0=`INLINE` (`# comment`)
+* 1=`EMBEDDED_DOCUMENT` (`=begin`/`=end`)
+* 2=`__END__` (after `__END__`)
+
+| # bytes | field |
+| --- | --- |
+| varint | comment type |
```

Comment:
Can this be a single byte? Since it's just 3 types and extremely unlikely to change.

Reply:
In practice it only takes one byte with `varint`.
I'd rather keep it simple here since there is AFAIK no real advantage to be a single byte, and it would add complexity/inconsistency in deserialization.

Reply:
Easy to change, done.